sale_commission_product_criteria_domain/models/commission_group.py

```diff
@@ -0,0 +1,70 @@
+# © 2023 ooops404
+# License AGPL-3 - See https://www.gnu.org/licenses/agpl-3.0.html
+from odoo import _, api, exceptions, fields, models
+
+
+class CommissionItemsGroup(models.Model):
+    _name = "commission.items.group"
+    _description = "Commission Items Group"
+    _sql_constraints = [
+        (
+            "unique_cig_name",
+            "UNIQUE(name)",
+            "Commission items group with such name already exists. "
+            "Name must be unique.",
+        )
+    ]
+
+    name = fields.Char(required=True)
+    commission_ids = fields.Many2many(
+        "sale.commission",
+        compute="_compute_commission_ids",
+        domain=[("commission_type", "=", "product_restricted")],
+        readonly=True,
+        store=True,
+    )
+    item_ids = fields.One2many(
+        "commission.item", "group_id", string="Items", readonly=True
+    )
+    agents_count = fields.Integer(compute="_compute_agents_count")
+
+    @api.depends("item_ids")
+    def _compute_commission_ids(self):
+        for rec in self:
+            rec.commission_ids = [(6, 0, rec.item_ids.mapped("commission_id").ids)]
+
+    def unlink(self):
+        if len(self.item_ids) > 0:
+            raise exceptions.ValidationError(
+                _(
+                    "You can not delete this commission group since "
+                    "there is related to it commission items."
+                )
+            )
+        return super().unlink()
+
+    def _compute_agents_count(self):
+        for rec in self:
+            agent_ids = self.env["res.partner"].search(
```

Comment:
Take out `self.env["res.partner"] from cycle.
Just set variable `res_partner_obj = self.env["res.partner"] before cycle.
And use `search_count()` for get records count and set result in record field.
`self.agents_count = res_partner_obj.search_count(...)`